calico/felix/labels.py

```diff
@@ -0,0 +1,480 @@
+
+    def _update_matches(self, expr_id, expr, item_id, label_values):
+        """
+        (Re-)evaluates the given expression against the given labels and
+        stores the result.
+        """
+        _log.debug("Re-evaluating %s against %s (%s)", expr_id, item_id,
+                   label_values)
+        if expr is not None and label_values is not None:
+            now_matches = expr.evaluate(label_values)
+            self.total_evaluations += 1
+            _log.debug("After evaluation, now matches: %s", now_matches)
+        else:
+            _log.debug("Expr or labels missing: no match")
+            now_matches = False
+        # Update the index and generate events.  These methods are idempotent
+        # so they'll ignore duplicate updates.
+        if now_matches:
+            self._store_match(expr_id, item_id)
+        else:
+            self._discard_match(expr_id, item_id)
+
+    def _store_match(self, expr_id, item_id):
+        """
+        Stores that an expression matches an item.
+
+        Calls on_match_started() as a side-effect. Idempotent, does
+        nothing if the match is already recorded.
+        """
+        previously_matched = self.matches_by_expr_id.contains(expr_id, item_id)
+        if not previously_matched:
+            _log.debug("%s now matches: %s", expr_id, item_id)
+            self.matches_by_expr_id.add(expr_id, item_id)
+            self.matches_by_item_id.add(item_id, expr_id)
+            self.on_match_started(expr_id, item_id)
+
+    def _discard_match(self, expr_id, item_id):
+        """
+        Stores that an expression does not match an item.
+
+        Calls on_match_stopped() as a side-effect.  Idempotent, does
+        nothing if the non-match is already recorded.
+        """
+        previously_matched = self.matches_by_expr_id.contains(expr_id, item_id)
+        if previously_matched:
+            _log.debug("%s no-longer matches %s", expr_id, item_id)
+            self.matches_by_expr_id.discard(expr_id, item_id)
+            self.matches_by_item_id.discard(item_id, expr_id)
+            self.on_match_stopped(expr_id, item_id)
+
+    def on_match_started(self, expr_id, item_id):
+        """
+        Called when an expression starts matching a particular set of
+        labels.
+
+        Intended to be assigned/overriden.
+        """
+        _log.debug("SelectorExpression %s now matches item %s",
+                   expr_id, item_id)
+
+    def on_match_stopped(self, expr_id, item_id):
+        """
+        Called when an expression stops matching a particular set of
+        labels.
+
+        Intended to be assigned/overriden.
+        """
+        _log.debug("SelectorExpression %s no longer matches item %s",
+                   expr_id, item_id)
+
+
+class LabelValueIndex(LinearScanLabelIndex):
+    """
+    Label index that indexes the values of labels, allowing for efficient
+    calculation of selectors of the form 'a == "b" && c == "d"', which
```

Comment:
Can you be clearer about what 'calculation' means here?  In particular, I think it doesn't mean 'evaluate', does it?